java-squid/src/main/java/org/sonar/java/se/AbstractBinaryConstraint.java

```diff
@@ -0,0 +1,119 @@
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.se;
+
+import org.sonar.java.se.ConstraintManager.BooleanConstraint;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+public abstract class AbstractBinaryConstraint {
+
+  public static AbstractBinaryConstraint contraintBetween(SymbolicValue v1, SymbolicValue v2, Collection<AbstractBinaryConstraint> constraints) {
+    for (AbstractBinaryConstraint constraint : constraints) {
+      AbstractBinaryConstraint c = constraint.getConstraintBetween(v1, v2, constraints);
+      if (c != null) {
+        return c;
+      }
+    }
+    return null;
+  }
+
+  protected final SymbolicValue v1;
+  protected final SymbolicValue v2;
+
+  protected AbstractBinaryConstraint(SymbolicValue v1, SymbolicValue v2) {
+    this.v1 = v1;
+    this.v2 = v2;
+  }
+
+  @Override
+  public boolean equals(Object obj) {
+    if (getClass().equals(obj.getClass())) {
+      return v1.equals(((AbstractBinaryConstraint) obj).v1) && v2.equals(((AbstractBinaryConstraint) obj).v2);
+    }
+    return false;
+  }
+
+  @Override
+  public int hashCode() {
+    long bits = v1.hashCode();
+    bits ^= 31 * v2.hashCode();
+    bits ^= 31 * getClass().hashCode();
+    return ((int) bits) ^ (int) (bits >> 32);
+  }
+
+  @Override
+  public String toString() {
+    StringBuilder buffer = new StringBuilder();
+    buffer.append('(');
+    buffer.append(v1);
+    buffer.append(getOperand());
+    buffer.append(v2);
+    buffer.append(')');
+    return buffer.toString();
+  }
+
+  protected AbstractBinaryConstraint getConstraintBetween(SymbolicValue v1, SymbolicValue v2, Collection<AbstractBinaryConstraint> constraints) {
+    AbstractBinaryConstraint c = null;
+    if (this.v1 == v1) {
+      if (this.v2 == v2) {
+        return this;
+      }
+      c = constraintUpTo(v2, withoutSelf(constraints));
+    } else if (this.v2 == v1) {
+      AbstractBinaryConstraint inverted = inverted();
+      if (inverted != null) {
+        if (this.v1 == v2) {
+          return inverted();
+        }
+        c = inverted().constraintUpTo(v2, withoutSelf(constraints));
+      }
+    }
+    return c;
+  }
+
+  private AbstractBinaryConstraint constraintUpTo(SymbolicValue sv2, Collection<AbstractBinaryConstraint> constraints) {
+    for (AbstractBinaryConstraint constraint : constraints) {
+      AbstractBinaryConstraint c = merge(constraint.getConstraintBetween(this.v2, sv2, constraints));
+      if (c != null) {
+        return c;
+      }
+    }
+    return null;
+  }
+
+  private Collection<AbstractBinaryConstraint> withoutSelf(Collection<AbstractBinaryConstraint> constraints) {
+    List<AbstractBinaryConstraint> constraintsList = new ArrayList<>(constraints);
+    constraintsList.remove(this); // Consume used constraint to prevent endless loop
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Move this trailing comment on the previous empty line. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3ATrailingCommentCheck)